src/crypto/DeviceList.js

```diff
@@ -58,31 +60,143 @@ const TRACKING_STATUS_UP_TO_DATE = 3;
+                        this._deviceTrackingStatus = (
+                            this._sessionStore.getEndToEndDeviceTrackingStatus() || {}
+                        );
+                        this._syncToken = this._sessionStore.getEndToEndDeviceSyncToken();
+                        this._cryptoStore.storeEndToEndDeviceData({
+                            devices: this._devices,
+                            trackingStatus: this._deviceTrackingStatus,
+                            syncToken: this._syncToken,
+                        }, txn);
+                        shouldDeleteSessionStore = true;
+                    } else {
+                        this._devices = deviceData ? deviceData.devices : {},
+                        this._deviceTrackingStatus = deviceData ?
+                            deviceData.trackingStatus : {};
+                        this._syncToken = deviceData ? deviceData.syncToken : null;
+                    }
+                });
+            },
+        );
+
+        if (shouldDeleteSessionStore) {
+            // migrated data is now safely persisted: remove from old store
+            this._sessionStore.removeEndToEndDeviceData();
+        }
+
         for (const u of Object.keys(this._deviceTrackingStatus)) {
             // if a download was in progress when we got shut down, it isn't any more.
             if (this._deviceTrackingStatus[u] == TRACKING_STATUS_DOWNLOAD_IN_PROGRESS) {
                 this._deviceTrackingStatus[u] = TRACKING_STATUS_PENDING_DOWNLOAD;
             }
         }
+    }
 
-        // userId -> promise
-        this._keyDownloadsInProgressByUser = {};
+    /**
+     * Save the device tracking state to storage, if any changes are
+     * pending other than updating the sync token
+     *
+     * @return {Promise<bool>} true is the data was saved, false if
+     *     it was not (eg. because no changes were pending)
+     */
+    async saveIfDirty() {
+        if (!this._dirty) return Promise.resolve(false);
+
+        if (this._savePromise === null) {
+            this._savePromise = Promise.delay(500).then(() => {
+                console.log('Saving device tracking data at token ' + this._syncToken);
+                return this._cryptoStore.doTxn(
+                    'readwrite', [IndexedDBCryptoStore.STORE_DEVICE_DATA], (txn) => {
+                        this._cryptoStore.storeEndToEndDeviceData({
+                            devices: this._devices,
+                            trackingStatus: this._deviceTrackingStatus,
+                            syncToken: this._syncToken,
+                        }, txn);
+                    },
+                );
+            }).then(() => {
+                this._dirty = false;
+                this._savePromise = null;
+                return true;
+            });
+        } else {
+            return this._savePromise;
+        }
+    }
+
+    /**
+     * Gets the current sync token
```

Comment:
the *current* sync token? I'm not sure I even know what it means for a sync token to be current.